core/iwasm/include/mem_alloc_export.h

```diff
@@ -0,0 +1,40 @@
+#ifndef _MEM_ALLOC_EXPORT_H_
+#define _MEM_ALLOC_EXPORT_H_
+
+#include <stdint.h>
+
+/* Memory allocator type */
+typedef enum {
+    /* pool mode, allocate memory from user defined heap buffer */
+    Alloc_With_Pool = 0,
+    /* user allocator mode, allocate memory from user defined
+       malloc function */
+    Alloc_With_Allocator,
+    /* system allocator mode, allocate memory from system allocator,
+       or, platform's os_malloc function */
+    Alloc_With_System_Allocator,
+    /* user allocator mode, allocate memory from user defined
+       malloc function with user data */
+    Alloc_With_Allocator_With_Data,
+} mem_alloc_type_t;
+
+/* Memory allocator option */
+typedef union MemAllocOption {
+    struct {
+        void *heap_buf;
+        uint32_t heap_size;
+    } pool;
+    struct {
+        void *malloc_func;
+        void *realloc_func;
+        void *free_func;
+    } allocator;
+    struct {
+        void *malloc_func;
+        void *realloc_func;
+        void *free_func;
+        void *data;
+    } allocator_with_data;
```

Comment:
I'm curious if you also explored alternative approach where we have a single struct (`allocator`) with `void* data` inside `ifdef`? It **might** make the code less readable, but that **possibly** would simplify the implementation? That's just an idea though, not sure if you already thought about it.

Reply:
Yes, I meant the build option, something like:
```c
typedef union MemAllocOption {
    struct {
        void *heap_buf;
        uint32_t heap_size;
    } pool;
    struct {
        void *malloc_func;
        void *realloc_func;
        void *free_func;
#ifdef MEM_ALLOC_WITH_USER_DATA
        void *user_data;
#endif
    } allocator;
} MemAllocOption;
```

Reply:
Good point, didn't realize that.